libjpeg_turbo: Add mingw support

seanybaggins opened the pull request:
## Description of changes
Needed for https://github.com/NixOS/nixpkgs/issues/272538
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingw32.libjpeg-turbo

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingwW64.libjpeg-turbo

seanybaggins commented:
@wegank can I snag a review?

wegank commented:
`libjpeg-turbo` -> `libjpeg_turbo` in the commit message, otherwise LGTM.

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingw32.libjpeg_turbo

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingwW64.libjpeg_turbo

Artturin reviewed (changes requested):
Please explain the reason why it's needed in the commit message and if the fix can be upstreamed.

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingw32.libjpeg_turbo

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingwW64.libjpeg_turbo

seanybaggins commented:
@Artturin addressed your comment

seanybaggins commented:
@Artturin originally got this working on 23.11. See https://github.com/seanybaggins/qt-package-overalys/tree/342f39a168779748870e89a53e8d54c67255d33c so they can be upstreamed to that release.

wegank commented on pkgs/development/libraries/libjpeg-turbo/default.nix:
```diff
@@ -42,9 +42,13 @@ stdenv.mkDerivation (finalAttrs: {
+      ./0001-Compile-transupp.c-as-part-of-the-library.patch
+    ] ++ lib.optionals stdenv.hostPlatform.isMinGW [
+      ./mingw-boolean.patch
+      ./transup-as-part-without-attribute-weak.patch
```
Can this be done via a `fetchPatch` from https://github.com/libjpeg-turbo/libjpeg-turbo/commit/537b9c336e8ba7f128b4171388b5d1b5b39f53ea? It would also be helpful to include the link to https://github.com/libjpeg-turbo/libjpeg-turbo/pull/487.

wegank commented:
> @Artturin addressed your comment

I'd prefer to keep `libjpeg-turbo: Add mingw support` as the title of commit message and put `Part of a greater effort ...` in another line.

seanybaggins committed:
libjpeg_turbo: add mingw support
Part of a greater effort to get qtbase and qtmultimedia cross compiling for
windows. This commit helps the qtbase cross compilation effor by fixing an
error when building the libjpeg_turbo reverse dependency.

seanybaggins commented:
@Artturin applied @wegank suggestion

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingw32.libjpeg_turbo

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingwW64.libjpeg_turbo

wegank reviewed (changes requested):
Oh, I just saw `libjpeg_turbo` bumped in #285136, the patch here won't apply.

RaitoBezarius commented on pkgs/development/libraries/libjpeg-turbo/default.nix:
```diff
@@ -10,7 +10,7 @@
 , enableShared ? !stdenv.hostPlatform.isStatic
 
-# for passthru.tests
+  # for passthru.tests
```
not needed reformatting

RaitoBezarius commented on pkgs/development/libraries/libjpeg-turbo/default.nix:
```diff
@@ -42,9 +42,13 @@ stdenv.mkDerivation (finalAttrs: {
+    lib.optionals (!stdenv.hostPlatform.isMinGW) [
+      ./0001-Compile-transupp.c-as-part-of-the-library.patch
+    ] ++ lib.optionals stdenv.hostPlatform.isMinGW [
+      ./mingw-boolean.patch
```
missing patch?

RaitoBezarius commented on pkgs/development/libraries/libjpeg-turbo/default.nix:
```diff
@@ -42,9 +42,13 @@ stdenv.mkDerivation (finalAttrs: {
-    ./mingw-boolean.patch;
+  patches =
+    lib.optionals (!stdenv.hostPlatform.isMinGW) [
+      ./0001-Compile-transupp.c-as-part-of-the-library.patch
```
missing patch?

RaitoBezarius reviewed (changes requested):
LGTM, in general, it's good practice to mention the provenance of patches, where did you find, what have you tried to upstream it, etc.

seanybaggins commented:
@wegank the change you mentioned was pushed into staging. Should I cherrypick from staging to incorporate those changes?

wegank commented:
That PR went to staging because it has too many rebuilds to land on master, so we certainly can't cherry-pick it for the same reason. The patches aren't the same for `libjpeg_turbo` 3.0.2 and 2.1.5.1, either.

Instead, if #285661, of which you are a co-author, passes the reviews and lands on staging, I think we can just close this PR and wait two or three weeks (or probably one, if you create PRs against staging-next).

seanybaggins commented:
Awesome. I will keep an eye on that PR. Appreciate your help in getting this across the finish line.

wegank commented:
Closed by #285661.
